## Data Wrangling

### **Gathering Data**


In [2]:
import pandas as pd

In [3]:
customer_df = pd.read_csv('https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/refs/heads/main/DicodingCollection/customers.csv')

customer_df.head()

customer_id customer_name             gender  age  \
0            1       fulan 1             Female   30   
1            2       fulan 2  Prefer not to say   69   
2            3       fulan 3  Prefer not to say   59   
3            4       fulan 4  Prefer not to say   67   
4            5       fulan 5  Prefer not to say   30   

                     home_address  zip_code               city  \
0  8606 Victoria TerraceSuite 560      5464      Johnstonhaven   
1      8327 Kirlin SummitApt. 461      8223    New Zacharyfort   
2       269 Gemma SummitSuite 109      5661           Aliburgh   
3       743 Bailey GroveSuite 141      1729  South Justinhaven   
4         48 Hyatt ManorSuite 375      4032     Griffithsshire   

                          state    country  
0            Northern Territory  Australia  
1               South Australia  Australia  
2  Australian Capital Territory  Australia  
3                    Queensland  Australia  
4                    Queensland  Australia

In [4]:
order_df = pd.read_csv('https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/refs/heads/main/DicodingCollection/orders.csv');

order_df.head()

order_id  customer_id  payment order_date delivery_date
0         1           64    30811  2021-8-30    2021-09-24
1         2          473    50490   2021-2-3    2021-02-13
2         3          774    46763  2021-10-8    2021-11-03
3         4          433    39782   2021-5-6    2021-05-19
4         5          441    14719  2021-3-23    2021-03-24

In [5]:
product_df = pd.read_csv('https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/refs/heads/main/DicodingCollection/products.csv')

product_df.head()

product_id product_type  product_name size colour  price  quantity  \
0           0        Shirt  Oxford Cloth   XS    red    114        66   
1           1        Shirt  Oxford Cloth    S    red    114        53   
2           2        Shirt  Oxford Cloth    M    red    114        54   
3           3        Shirt  Oxford Cloth    L    red    114        69   
4           4        Shirt  Oxford Cloth   XL    red    114        47   

                                    description  
0  A red coloured, XS sized, Oxford Cloth Shirt  
1   A red coloured, S sized, Oxford Cloth Shirt  
2   A red coloured, M sized, Oxford Cloth Shirt  
3   A red coloured, L sized, Oxford Cloth Shirt  
4  A red coloured, XL sized, Oxford Cloth Shirt

In [6]:
sales_df = pd.read_csv('https://raw.githubusercontent.com/dicodingacademy/dicoding_dataset/refs/heads/main/DicodingCollection/sales.csv')

sales_df.head()

sales_id  order_id  product_id  price_per_unit  quantity  total_price
0         0         1         218             106         2        212.0
1         1         1         481             118         1        118.0
2         2         1           2              96         3        288.0
3         3         1        1002             106         2        212.0
4         4         1         691             113         3        339.0

**Assessing Data and Cleaning Data**
### Menilai Customer_df

In [7]:
customer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1007 entries, 0 to 1006
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   customer_id    1007 non-null   int64 
 1   customer_name  1007 non-null   object
 2   gender         989 non-null    object
 3   age            1007 non-null   int64 
 4   home_address   1007 non-null   object
 5   zip_code       1007 non-null   int64 
 6   city           1007 non-null   object
 7   state          1007 non-null   object
 8   country        1007 non-null   object
dtypes: int64(3), object(6)
memory usage: 70.9+ KB


In [8]:
customer_df.isna().sum()

customer_id       0
customer_name     0
gender           18
age               0
home_address      0
zip_code          0
city              0
state             0
country           0
dtype: int64

In [9]:
print("Jumlah duplikasi : ", customer_df.duplicated().sum())

Jumlah duplikasi :  6


In [10]:
customer_df.describe()

customer_id          age     zip_code
count  1007.000000  1007.000000  1007.000000
mean    501.726912    50.929494  5012.538232
std     288.673238    30.516299  2885.836112
min       1.000000    20.000000     2.000000
25%     252.500000    34.000000  2403.500000
50%     502.000000    50.000000  5087.000000
75%     751.500000    65.000000  7493.500000
max    1000.000000   700.000000  9998.000000

#### Hasil Penilaian / Assesment customer_df
Masalah yang ada pada **customer_df**

*   18 data pada kolom gender yang NULL / Kosong
*   6 buah data duplikasi
*   kolom umur yang tidak akurat, dimana umur maksimal 700



### Membersihkan customer_df

#### Menghapus data yang duplikat


In [11]:
# Menghapus data yang duplikat
customer_df.drop_duplicates(inplace=True)

print('Jumlah duplikasi :', customer_df.duplicated().sum())

Jumlah duplikasi : 0


#### Mengisi Missing Value

In [12]:
# Mengambil data yang memiliki missing value
customer_df[customer_df.gender.isna()]

customer_id customer_name gender  age                     home_address  \
38            39      fulan 39    NaN   80  7440 Cameron Estate DrSuite 628   
167          168     fulan 168    NaN   27         2781 Berge MallSuite 452   
322          322     fulan 322    NaN   30        593 Becker CircleApt. 333   
393          393     fulan 393    NaN   34          5158 Levi HillSuite 531   
442          442     fulan 442    NaN   26           5157 Feil RoadApt. 633   
722          720     fulan 720    NaN   40         31 Jordan ParadeApt. 400   
745          743     fulan 743    NaN   57   09 Christopher StreetSuite 967   
773          771     fulan 771    NaN   74     7367 Wright JunctionApt. 773   
798          795     fulan 795    NaN   49          487 Summer MewsApt. 874   
801          798     fulan 798    NaN   56           27 Aiden KnollApt. 875   
825          822     fulan 822    NaN   59        41 Jenkins KnollSuite 438   
859          855     fulan 855    NaN   55       603 O'keefe KnollSuite 782   
863          859     fulan 859    NaN   38            32 Isla GroveApt. 078   
914          909     fulan 909    NaN   62    976 Murray Station StApt. 036   
934          929     fulan 929    NaN   68           394 Lily HillSuite 153   
948          943     fulan 943    NaN   64       3117 Heller PlaceSuite 149   
952          946     fulan 946    NaN   24      8227 Nicholas HillSuite 150   
994          988     fulan 988    NaN   35   1130 Turner Estate DrSuite 925   

     zip_code                   city                         state    country  
38       4622  North Victoriachester            Northern Territory  Australia  
167      1975         North Leoburgh             Western Australia  Australia  
322      1640             Jacobiview             Western Australia  Australia  
393      1474             Johnsburgh                    Queensland  Australia  
442      7249             Port Chloe               New South Wales  Australia  
722      1380             West Henry               South Australia  Australia  
745      6226         Lake Lukemouth             Western Australia  Australia  
773      8882               Kuhntown                      Victoria  Australia  
798      1712            East Hayden  Australian Capital Territory  Australia  
801      6531               Port Sam  Australian Capital Territory  Australia  
825      2588        Lake Andrewport               South Australia  Australia  
859      8822        Port Dylanmouth                      Tasmania  Australia  
863      7711            Rosechester               New South Wales  Australia  
914      3227               Langfort                      Tasmania  Australia  
934      2353             Beahanfurt            Northern Territory  Australia  
948       822           North Elijah               South Australia  Australia  
952       115           South Jasper                    Queensland  Australia  
994      9386              New Harry             Western Australia  Australia

Berdasarkan gambar di atas, dapat dilihat bahwa baris data tersebut masih mengandung banyak informasi penting sehingga sayang jika langsung dibuang. Oleh karena itu, pada kasus ini, kita akan menggunakan metode imputation untuk menangani missing value.

In [13]:
# Gunakanlah method value_counts()untuk mengidentifikasi nilai yang dominan.
customer_df.gender.value_counts()

gender
Prefer not to say    725
Male                 143
Female               115
Name: count, dtype: int64

Dominan data pada kategori gender yaitu Prefer not to say. Nilai inilah yang selanjutnya akan kita gunakan sebagai pengganti missing value. Proses penggantian ini dapat dilakukan menggunakan method fillna()

In [14]:
customer_df.fillna(value='Prefer not to say', inplace=True)

# cek apakah masih ada kolom data yang NaN
customer_df.isna().sum()

customer_id      0
customer_name    0
gender           0
age              0
home_address     0
zip_code         0
city             0
state            0
country          0
dtype: int64

#### Mencari inaccurate data

In [15]:
customer_df[customer_df.age == customer_df.age.max()]

customer_id customer_name             gender  age  \
967          961     fulan 961  Prefer not to say  700   

                   home_address  zip_code        city            state  \
967  29 Farrell ParadeSuite 818      6528  New Joseph  South Australia   

       country  
967  Australia

kita bisa berasumsi bahwa inaccurate value tersebut terjadi karena human error sehingga kelebihan memasukkan nilai nol. Oleh karena itu, gantilah dengan nilai 70. Proses ini dilakukan dengan memanfaatkan method replace()

In [16]:
customer_df.replace({customer_df['age'].max(): 70}, inplace=True)

In [17]:
customer_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    500.312687    50.323676  5000.693307
std     289.265537    22.665946  2886.084454
min       1.000000    20.000000     2.000000
25%     250.000000    34.000000  2398.000000
50%     500.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000   500.000000  9998.000000

In [18]:
customer_df.replace({customer_df['age'].max(): 50}, inplace=True)

In [19]:
customer_df.describe()

customer_id          age     zip_code
count  1001.000000  1001.000000  1001.000000
mean    499.863137    49.874126  5000.693307
std     289.615487    17.644663  2886.084454
min       1.000000    20.000000     2.000000
25%     249.000000    34.000000  2398.000000
50%     499.000000    50.000000  5079.000000
75%     751.000000    65.000000  7454.000000
max    1000.000000    80.000000  9998.000000

###Menilai Data orders_df


In [20]:
order_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   order_id       1000 non-null   int64 
 1   customer_id    1000 non-null   int64 
 2   payment        1000 non-null   int64 
 3   order_date     1000 non-null   object
 4   delivery_date  1000 non-null   object
dtypes: int64(3), object(2)
memory usage: 39.2+ KB


In [21]:
order_df.isna().sum()

order_id         0
customer_id      0
payment          0
order_date       0
delivery_date    0
dtype: int64

In [22]:
print("Jumlah duplikasi: ",order_df.duplicated().sum())
order_df.describe()

Jumlah duplikasi:  0


order_id  customer_id       payment
count  1000.000000  1000.000000   1000.000000
mean    500.500000   506.640000  33972.936000
std     288.819436   277.115502  14451.609047
min       1.000000     1.000000  10043.000000
25%     250.750000   275.250000  21329.250000
50%     500.500000   515.000000  33697.500000
75%     750.250000   737.250000  46249.000000
max    1000.000000  1000.000000  59910.000000

Masalah yang ada pada **order_df**

*   Kolom order_date dan delivery_date menggunakan tipe data object / string, seharusnya datetime


### Membersihkan order_df

#### Mengganti tipe data dari Object menjadi datetime

Untuk mengatasi masalah ini, kita akan mengganti tipe data pada kolom order_date & delivery_date menjadi datetime. Proses ini dapat dilakukan menggunakan function to_datetime() yang disediakan oleh library pandas

In [23]:
datetime_columns = ["order_date", "delivery_date"]

for column in datetime_columns:
  order_df[column] = pd.to_datetime(order_df[column])

In [24]:
order_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   order_id       1000 non-null   int64         
 1   customer_id    1000 non-null   int64         
 2   payment        1000 non-null   int64         
 3   order_date     1000 non-null   datetime64[ns]
 4   delivery_date  1000 non-null   datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 39.2 KB


### Menilai product_df

In [25]:
product_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1266 entries, 0 to 1265
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   product_id    1266 non-null   int64 
 1   product_type  1266 non-null   object
 2   product_name  1266 non-null   object
 3   size          1266 non-null   object
 4   colour        1266 non-null   object
 5   price         1266 non-null   int64 
 6   quantity      1266 non-null   int64 
 7   description   1266 non-null   object
dtypes: int64(3), object(5)
memory usage: 79.2+ KB


In [26]:
print("Jumlah duplikasi: ",product_df.duplicated().sum())
product_df.describe()

Jumlah duplikasi:  6


product_id        price     quantity
count  1266.000000  1266.000000  1266.000000
mean    627.926540   105.812006    60.138231
std     363.971586     9.715611    11.682791
min       0.000000    90.000000    40.000000
25%     313.250000    95.250000    50.000000
50%     626.500000   109.000000    60.000000
75%     942.750000   114.000000    70.000000
max    1259.000000   119.000000    80.000000

Masalah yang ada pada product_df

*  Terdapat 6 buah data yang terindikasi duplikasi

### Membersihkan product_df

#### Menghapus data yang duplikat / sama

In [27]:
product_df.drop_duplicates(inplace=True)

In [28]:
print("Jumlah duplikasi: ", product_df.duplicated().sum())

Jumlah duplikasi:  0


### Menilai Sales_df

In [29]:
sales_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   sales_id        5000 non-null   int64  
 1   order_id        5000 non-null   int64  
 2   product_id      5000 non-null   int64  
 3   price_per_unit  5000 non-null   int64  
 4   quantity        5000 non-null   int64  
 5   total_price     4981 non-null   float64
dtypes: float64(1), int64(5)
memory usage: 234.5 KB


In [30]:
sales_df.isna().sum()

sales_id           0
order_id           0
product_id         0
price_per_unit     0
quantity           0
total_price       19
dtype: int64

In [31]:
print("Jumlah duplikasi: ",sales_df.duplicated().sum())
sales_df.describe()

Jumlah duplikasi:  0


sales_id     order_id   product_id  price_per_unit    quantity  \
count  5000.000000  5000.000000  5000.000000     5000.000000  5000.00000   
mean   2499.500000   503.038200   634.053200      103.501600     1.99240   
std    1443.520003   285.964418   363.255794        9.195004     0.80751   
min       0.000000     1.000000     1.000000       90.000000     1.00000   
25%    1249.750000   258.000000   323.000000       95.000000     1.00000   
50%    2499.500000   504.500000   635.000000      102.000000     2.00000   
75%    3749.250000   749.000000   951.000000      112.000000     3.00000   
max    4999.000000   999.000000  1259.000000      119.000000     3.00000   

       total_price  
count  4981.000000  
mean    206.307368  
std      86.352449  
min      90.000000  
25%     112.000000  
50%     204.000000  
75%     285.000000  
max     357.000000

Masalah yang ada pada sales_df

*  19 data pada kolom total_price yang NULL / Kosong

### Membersihkan sales_df

#### Mengisi Missing Value

In [32]:
# Mencari data yang ada missing value
sales_df[sales_df.total_price.isna()]

sales_id  order_id  product_id  price_per_unit  quantity  total_price
9            9         2        1196             105         1          NaN
121        121        27        1027              90         3          NaN
278        278        63         360              94         2          NaN
421        421        95        1091             115         1          NaN
489        489       108        1193             105         3          NaN
539        539       117         405             119         2          NaN
636        636       134         653              93         3          NaN
687        687       145        1138             102         1          NaN
854        854       177          64             104         1          NaN
1079      1079       222         908              94         3          NaN
1193      1193       248        1121             102         2          NaN
1313      1313       272         826             117         1          NaN
1548      1548       316         103             118         3          NaN
1688      1688       345         428             107         1          NaN
1775      1775       359         694             113         2          NaN
1902      1902       381        1218             105         3          NaN
2025      2025       408         611             112         3          NaN
2164      2164       436         583             100         3          NaN
2347      2347       476         696             113         2          NaN

In [33]:
# Filling missing value of total price with multiply price_per_unit and
sales_df["total_price"] = sales_df["price_per_unit"] * sales_df["quantity"]

In [34]:
# Checking if there is any missing value
sales_df.isna().sum()

sales_id          0
order_id          0
product_id        0
price_per_unit    0
quantity          0
total_price       0
dtype: int64

## Exploratory Data Analysis

Berdasarkan beberapa masalah tersebut, kita dapat mendefinisikan beberapa pertanyaan bisnis seperti berikut.

*  Bagaimana performa penjualan dan revenue perusahaan dalam beberapa bulan terakhir?
*  Produk apa yang paling banyak dan paling sedikit terjual?
*  Bagaimana demografi pelanggan yang kita miliki?
*  Kapan terakhir pelanggan melakukan transaksi?
*  Seberapa sering seorang pelanggan melakukan pembelian dalam beberapa bulan terakhir?
*  Berapa banyak uang yang dihabiskan pelanggan dalam beberapa bulan terakhir?

### Eksplorasi Data customer_df

Untuk mendapat intisari ringkas / rangkuman dari data customer_df

In [35]:
customer_df.describe(include='all')

customer_id customer_name             gender          age  \
count   1001.000000          1001               1001  1001.000000   
unique          NaN          1000                  3          NaN   
top             NaN     fulan 943  Prefer not to say          NaN   
freq            NaN             2                743          NaN   
mean     499.863137           NaN                NaN    49.874126   
std      289.615487           NaN                NaN    17.644663   
min        1.000000           NaN                NaN    20.000000   
25%      249.000000           NaN                NaN    34.000000   
50%      499.000000           NaN                NaN    50.000000   
75%      751.000000           NaN                NaN    65.000000   
max     1000.000000           NaN                NaN    80.000000   

                      home_address     zip_code     city            state  \
count                         1001  1001.000000     1001             1001   
unique                        1000          NaN      961                8   
top     3117 Heller PlaceSuite 149          NaN  New Ava  South Australia   
freq                             2          NaN        3              140   
mean                           NaN  5000.693307      NaN              NaN   
std                            NaN  2886.084454      NaN              NaN   
min                            NaN     2.000000      NaN              NaN   
25%                            NaN  2398.000000      NaN              NaN   
50%                            NaN  5079.000000      NaN              NaN   
75%                            NaN  7454.000000      NaN              NaN   
max                            NaN  9998.000000      NaN              NaN   

          country  
count        1001  
unique          1  
top     Australia  
freq         1001  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

Berdasarkan rangkuman parameter statistik di atas, kita akan memperoleh informasi jumlah pelanggan sebanyak 1001 orang yang berumur di kisaran antara 20 hingga 80 tahun dengan rata-rata umur sebesar 49.87 tahun dengan standar deviasinya sebesar 17.64 tahun.

In [36]:
customer_df.groupby(by='gender').agg({
    'customer_id': 'nunique',
    'age': ['min', 'max', 'mean', "std"]
})

customer_id age                          
                      nunique min max       mean        std
gender                                                     
Female                    115  20  79  49.147826  16.646607
Male                      143  20  80  51.230769  18.462635
Prefer not to say         741  20  80  49.725437  17.644283

*   Berdasarkan pivot table di atas, dapat diketahui bahwa pelanggan yang kita miliki didominasi oleh jenis kelamin prefer not to say.
*   Di lain sisi, distribusi umur mereka
ternyata cukup mirip yaitu berkisar antara 20 hingga 80 tahun

In [37]:
customer_df.groupby(by='city').customer_id.nunique().sort_values(ascending=False)

city
East Aidan       3
East Sophia      3
New Ava          3
West Jackfort    2
Corkeryshire     2
                ..
Jordanside       1
Josephland       1
Josephmouth      1
Justinport       1
Zacville         1
Name: customer_id, Length: 961, dtype: int64

In [38]:
customer_df.groupby(by='state').customer_id.nunique().sort_values(ascending=False)

state
South Australia                 139
Queensland                      134
New South Wales                 132
Northern Territory              125
Western Australia               123
Australian Capital Territory    121
Victoria                        121
Tasmania                        104
Name: customer_id, dtype: int64

*  Berdasarkan hasil di atas, dapat dilihat bahwa persebaran pelanggan kita cukup merata pada setiap kota dan negara bagian.
*  Pelanggan kita paling banyak berada di kota East Aidan, East Sophia, dan New Ava dengan jumlah masing-masing tiga pelanggan.
*  Selain itu, pelanggan kita paling banyak berasal dari negara bagian South Australia.

### Eksplorasi Data order_df

In [39]:
delivery_time = order_df['delivery_date'] - order_df['order_date']
delivery_time = delivery_time.apply(lambda x: x.total_seconds())
order_df['delivery_time (days)'] = round(delivery_time/86400)

In [40]:
display(order_df)

order_id  customer_id  payment order_date delivery_date  \
0           1           64    30811 2021-08-30    2021-09-24   
1           2          473    50490 2021-02-03    2021-02-13   
2           3          774    46763 2021-10-08    2021-11-03   
3           4          433    39782 2021-05-06    2021-05-19   
4           5          441    14719 2021-03-23    2021-03-24   
..        ...          ...      ...        ...           ...   
995       996          345    37843 2021-01-13    2021-02-02   
996       997          346    53831 2021-01-18    2021-01-31   
997       998          407    53308 2021-05-05    2021-05-21   
998       999          428    31643 2021-06-15    2021-07-12   
999      1000          896    27836 2021-04-07    2021-04-24   

     delivery_time (days)  
0                    25.0  
1                    10.0  
2                    26.0  
3                    13.0  
4                     1.0  
..                    ...  
995                  20.0  
996                  13.0  
997                  16.0  
998                  27.0  
999                  17.0  

[1000 rows x 6 columns]

In [41]:
order_df.describe(include="all")

order_id  customer_id       payment                  order_date  \
count  1000.000000  1000.000000   1000.000000                        1000   
mean    500.500000   506.640000  33972.936000  2021-05-27 18:38:52.800000   
min       1.000000     1.000000  10043.000000         2021-01-01 00:00:00   
25%     250.750000   275.250000  21329.250000         2021-03-13 18:00:00   
50%     500.500000   515.000000  33697.500000         2021-05-27 12:00:00   
75%     750.250000   737.250000  46249.000000         2021-08-12 06:00:00   
max    1000.000000  1000.000000  59910.000000         2021-10-24 00:00:00   
std     288.819436   277.115502  14451.609047                         NaN   

             delivery_date  delivery_time (days)  
count                 1000           1000.000000  
mean   2021-06-10 20:31:12             14.078000  
min    2021-01-03 00:00:00              1.000000  
25%    2021-03-28 00:00:00              8.000000  
50%    2021-06-11 12:00:00             14.000000  
75%    2021-08-24 06:00:00             21.000000  
max    2021-11-20 00:00:00             27.000000  
std                    NaN              7.707225

### Eksplorasi Data order_df dan customer_df

Buat kolom baru bernama Status, Aktif ketika customer_id terdeteksi pada kolom order_df dan sebaliknya

In [42]:
customer_id_in_order_df = order_df.customer_id.tolist()
customer_df["status"] = customer_df["customer_id"].apply(lambda x: "Active" if x in customer_id_in_order_df else "Non Active")
customer_df.sample(5)

customer_id customer_name             gender  age  \
247          248     fulan 248  Prefer not to say   38   
853          850     fulan 850               Male   55   
976          970     fulan 970  Prefer not to say   44   
601          600     fulan 600               Male   50   
677          676     fulan 676  Prefer not to say   59   

                      home_address  zip_code             city  \
247       00 Fadel CircuitApt. 573      4463      D'amorefort   
853       1488 Kuhn AvenueApt. 525      9923      Lillianside   
976         967 Skye DriveApt. 746      7005     Lake Lincoln   
601           41 Aiden WayApt. 249      2075  East Archieside   
677  140 Jessica BoulevardApt. 214      1938        Kaylafurt   

                            state    country      status  
247            Northern Territory  Australia  Non Active  
853               New South Wales  Australia  Non Active  
976                    Queensland  Australia      Active  
601  Australian Capital Territory  Australia  Non Active  
677            Northern Territory  Australia      Active

In [43]:
customer_df.groupby(by="status").customer_id.count()

status
Active        618
Non Active    383
Name: customer_id, dtype: int64

Seperti yang terlihat pada pivot table di atas, terdapat cukup banyak pelanggan yang belum pernah melakukan transaksi sebelumnya. Ini tentunya merupakan kabar buruk bagi kita karena hampir 30% pelanggan kita belum pernah melakukan order sebelumnya.

### Join tabel order_df dan customer_df dengan mode left join.

In [44]:
order_customer_df = pd.merge(
    left = order_df,
    right = customer_df,
    how = "left",
    left_on = "customer_id",
    right_on = "customer_id"
)
order_customer_df.head()

order_id  customer_id  payment order_date delivery_date  \
0         1           64    30811 2021-08-30    2021-09-24   
1         2          473    50490 2021-02-03    2021-02-13   
2         3          774    46763 2021-10-08    2021-11-03   
3         4          433    39782 2021-05-06    2021-05-19   
4         5          441    14719 2021-03-23    2021-03-24   

   delivery_time (days) customer_name             gender   age  \
0                  25.0      fulan 64  Prefer not to say  75.0   
1                  10.0     fulan 473               Male  61.0   
2                  26.0     fulan 774  Prefer not to say  34.0   
3                  13.0     fulan 433  Prefer not to say  46.0   
4                   1.0     fulan 441  Prefer not to say  53.0   

                    home_address  zip_code                city  \
0      4927 Alice MeadowApt. 960    7787.0      Sanfordborough   
1  531 Schmitt BoulevardApt. 010    1744.0             Annaton   
2       2096 Wilson MewsApt. 714    8590.0  West Jonathanshire   
3        5777 Mayer PassApt. 881    9728.0    West Michaelport   
4   33 Richards JunctionApt. 478    7650.0     South Rileyview   

               state    country  status  
0    South Australia  Australia  Active  
1    South Australia  Australia  Active  
2           Tasmania  Australia  Active  
3           Tasmania  Australia  Active  
4  Western Australia  Australia  Active

### Jumlah Order berdasarkan kota

In [45]:
order_customer_df.groupby(by="city").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

city  order_id
0        Jordanside         6
1           New Ava         6
2          West Kai         5
3        O'keefeton         5
4          East Max         5
5         Lake Rose         5
6          Rubyfort         5
7  Port Hannahburgh         5
8         Kiehnfurt         4
9        Lake Jesse         4

berdasarkan hasil tersebut, diketahui bahwa kota Jordanside dan New Ava merupakan dua kota yang memiliki jumlah order terbanyak.

### Jumlah Order berdasarkan state

In [48]:
order_customer_df.groupby(by="state").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

state  order_id
0               South Australia       148
1                    Queensland       139
2               New South Wales       130
3             Western Australia       129
4  Australian Capital Territory       118
5                      Victoria       118
6                      Tasmania       112
7            Northern Territory       106

South Australia merupakan state yang paling banyak melakukan order

### Jumlah order berdasarkan gender

In [49]:
order_customer_df.groupby(by="gender").order_id.nunique().sort_values(ascending=False).reset_index().head(10)

gender  order_id
0  Prefer not to say       725
1             Female       139
2               Male       136

kebanyakan order dilakukan oleh pelanggan yang berjenis kelamin prefer not to say

### Jumlah order berdasarkan kelompok usia

In [50]:
order_customer_df["age_group"] = order_customer_df.age.apply(lambda x: "Youth" if x <= 24 else ("Seniors" if x > 64 else "Adults"))
order_customer_df.groupby(by="age_group").order_id.nunique().sort_values(ascending=False)

age_group
Adults     682
Seniors    227
Youth       93
Name: order_id, dtype: int64

pelanggan yang paling banyak melakukan order berasal dari kelompok usia Adults.